src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java

```diff
@@ -81,67 +84,91 @@
     private static final long serialVersionUID = 1574325691L;
 
-    public static class ImageReferenceTypeClass {
-        private String image;
-        private String imageId;
-        private String imagePublisher;
-        private String imageOffer;
-        private String imageSku;
-        private String imageVersion;
+    public static class ImageReferenceTypeClass implements Serializable {
+        private String uri;
+        private String id;
+        private String publisher;
+        private String offer;
+        private String sku;
+        private String version;
         private String galleryName;
         private String galleryImageDefinition;
         private String galleryImageVersion;
         private String gallerySubscriptionId;
         private String galleryResourceGroup;
+        private ImageReferenceType type;
 
         @DataBoundConstructor
         public ImageReferenceTypeClass(
-                String image,
-                String imageId,
-                String imagePublisher,
-                String imageOffer,
-                String imageSku,
-                String imageVersion,
+                String uri,
+                String id,
+                String publisher,
+                String offer,
+                String sku,
+                String version,
                 String galleryName,
                 String galleryImageDefinition,
                 String galleryImageVersion,
                 String gallerySubscriptionId,
                 String galleryResourceGroup) {
-            this.image = image;
-            this.imageId = imageId;
-            this.imagePublisher = imagePublisher;
-            this.imageOffer = imageOffer;
-            this.imageSku = imageSku;
-            this.imageVersion = imageVersion;
+            this.uri = uri;
+            this.id = id;
+            this.publisher = publisher;
+            this.offer = offer;
+            this.sku = sku;
+            this.version = version;
             this.galleryName = galleryName;
             this.galleryImageDefinition = galleryImageDefinition;
             this.galleryImageVersion = galleryImageVersion;
             this.gallerySubscriptionId = gallerySubscriptionId;
             this.galleryResourceGroup = galleryResourceGroup;
+
+            this.type = determineType();
+        }
+
+        private ImageReferenceTypeClass() {
+            // used for readResolve to maintain compatibility
+        }
+
+        private ImageReferenceType determineType() {
+            if (Util.fixEmpty(uri) != null) {
+                return ImageReferenceType.CUSTOM;
+            }
+            if (Util.fixEmpty(id) != null) {
+                return ImageReferenceType.CUSTOM_IMAGE;
+            }
+            if (Util.fixEmpty(galleryName) != null) {
+                return ImageReferenceType.GALLERY;
+            }
+            return ImageReferenceType.REFERENCE;
```

Comment:
Here a template will get an existing referenceType anyway which may influence some scenarios.

Reply:
Isn't this how the code behaved before or did I miss something?
What are you suggesting to be changed?

Reply:
I just debug the previous version, when using the basic provided images, the image reference here will be `null` which may not be suitable. But now it will be `ImageReferenceType.REFERENCE` and affect some condition check.

Reply:
The `isType` method defaulted to returning true when it was reference
```
        if (this.imageReferenceType == null && type.equals("reference")) {
            return true;
        }
```
and the `getImageReferenceType` also defaulted to REFERENCE previously

As far as I can tell while it was previously `null` this is just always setting the correct default.

Reply:
In my mind, this plugin uses two variables to identify the configuration of images. One is `isBasic` and the other is `imageReferenceType`. When users choose the basic images, there will be no parameters for reference provided by users. In this case the `imageReferenceType` variable should not be set as `ImageReferenceType.REFERENCE` which using `imagePublisher`, `imageOffer`... to point to an image.

Reply:
I see, almost everything seems to be guarded by `isBasic` and the ones that aren't are looking at different types.